crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright 2015-2022 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// https://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// https://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{io, path::Path};
+
+use tracing::{debug, info, trace};
+
+use crate::{
+    authority::{
+        Authority, LookupError, LookupObject, LookupOptions, MessageRequest, UpdateResult, ZoneType,
+    },
+    proto::{
+        op::{Query, ResponseCode},
+        rr::{RData, LowerName, Name, Record, RecordType, rdata::A},
+    },
+    server::RequestInfo,
+    store::blocklist::BlockListConfig,
+};
+
+use crate::resolver::lookup::Lookup;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::prelude::*;
+
+/// An authority that will resolve queries against one or more block lists.  The typical use case will be to use this in a chained
+/// configuration before a forwarding or recursive resolver:
+///
+///   [[zones]]
+///   zone = "."
+///   zone_type = "hint"
+///   stores = [{ type = "blocklist", lists = ["default/bl.txt", "default/bl2.txt"]}, { type = "recursor", roots = "default/root.zone"}]
+///
+/// Note: the order of the stores is important: the first one specified in the store list will be the first consulted.  Subsequent stores
+/// will only be consulted if each prior store returns Err(LookupError::NotHandled) in response to the query.
+
+pub struct BlockListAuthority {
+    origin: LowerName,
+    block_table: HashMap<String,bool>, // String: key, bool: wildcard?
```

Comment:
Storing this as a String means an expensive conversion to_string on each call in the search. Also, the wildcard lookup looks expensive. I wonder if it might be better to do a BTreeMap with `LowerName` instead. LowerName already has a wildcard designation. Then on search you could take the LowerName and strip the labels until you find a matching label that is a Wildcard. You can see this logic here: https://github.com/hickory-dns/hickory-dns/blob/8f05d14eed53c488ee14fa860ef68cad352168ff/crates/server/src/store/in_memory/authority.rs#L415

Reply:
This is stored as a lowername, but I still have a few other string-related things to clean up.